### DISCUSSION TITLE
Modules - more control over options

### DISCUSSION THREAD

**Discussion by Qrzy** [nuxt/nuxt#15932] [Ideas]
Hi Everyone! 👋 

As I've recently learned 📖  in https://github.com/Baroshem/nuxt-security/issues/78, properly reading and applying module options happens to be tricky. 😃  That case particularly narrows down to `defu` concatenating arrays when we might want to replace them 🤯  and my proposed solution there is not to rely on Nuxt's mechanisms but merge the defaults and applied options in a customized way.

However, on second thought, I found out it's actually something that begs for an improvement in Nuxt itself. 🙏 
I'm not much into Nuxt's internal architecture, hence I need to emphasize here that I am not proposing any particular solution, but wanted to raise this for discussion. 💡 

Namely, in Nuxt Kit's code we've got this: ⚠️ 
https://github.com/nuxt/framework/blob/5ac9d85a497d6003cbe0c5bbf0e16ef6ea69b46a/packages/kit/src/module/define.ts#L26
and that variable is eventually returned and then provided to the module definition, as `options`.
This, especially in case of array values, leads to the situation where there is an arbitrary logic in Nuxt merging the options and as a result the module author does not have full control over those. 😞 
Even not considering this special case of `defu` vs arrays - there might be cases, where the module author might want to apply a custom logic when merging default options with those given by the module user. Currently, there is none given. And in case of arrays it's actually pretty hard to manage if we take into consideration things like in this minimal (I hope) reproduction repo:
https://github.com/Qrzy/nuxt-module-opts-repro
(`yarn && yarn dev` will run the app that will show the config in the page)

There we have a module, that just injects its options to the runtime config, but the playground app extends a nested nuxt.config. That shows what happens even without providing defaults in module definition:
- we get all the arrays concatenated - and that's arbitrary
- we're not able to make them override each other

❓ What do you think of giving the module author some control, or an _option_ to take control over options merging? 
May the simplest example be `defu`'s custom merger, but that's just first thing that pops up.

**Comment by TheAlexLichter** [comment]
Hey! Thanks for the detailed post!

* Defu already supports custom mergers based on functions (e.g. used for multi-layers, see https://github.com/nuxt/framework/issues/9812). They are *not* in use for Nuxt module though by default.
* There is also the `schema` key which allows to define a custom schema via [untyped](https://github.com/unjs/untyped/) that also handles default values and merging.

I haven't played around with the schema key much yet but it could be helpful in this scenario. The only possible issue is that it runs *after* the defaults are merged 🙈

If `schema` is not enough, a custom `defu` instance could be provided (not possible yet AFAIK) 🤔

**Reply by Qrzy** [reply]
Thanks for the reply, this provides some new things I should learn about. :) 

Where can I find more on `schema`? 
I can't see it in the [Module Author Guide](https://nuxt.com/docs/guide/going-further/modules).

**Reply by TheAlexLichter** [reply]
Sadly there is no documentation about using the `schema` key for modules so the best bet is the source code + https://untyped.unjs.io/ + [the repo](https://github.com/unjs/untyped/) 🙈

**Reply by Qrzy** [reply]
And re:
> a custom defu instance could be provided (not possible yet AFAIK)

that's the essence of my questions in the end of the first post, actually 😃

**Comment by Qrzy** [comment]
Is it allowed to bump, maybe? 😃

**Comment by jd1378** [comment]
this is a serious issue that needs to be addressed immediately imo

I had a recent contribution to `nuxt-svgo`, that I found out that one of module options array is being merged with the default config, which is a disaster depending on what you want to do. in our case a user was depending on svg dimensions to exist, default config had a plugin in the array to remove it, they didn't have change their config, I expected the user config's array to replace the default, instead they got merged. how is the user supposed to be able to remove a plugin that is in an array if It's only going to be merged.
